controllers/quay/features.go

```diff
@@ -35,267 +33,311 @@ const (
-
-		// TODO(alecmerdler): Try to fetch the wildcard cert from the `ConfigMap` at `openshift-config-managed/default-ingress-cert`...
-		clusterWildcardCert, err := getCertificatesPEM(fakeRoute.(*routev1.Route).Spec.Host + ":443")
-		if err != nil {
-			return ctx, quay, err
-		}
-
-		ctx.ClusterWildcardCert = clusterWildcardCert
-
-		r.Log.Info("detected cluster wildcard certificate for " + ctx.ClusterHostname)
+			if len(rt.Status.Ingress) == 0 {
+				r.Log.Info("waiting to detect `routerCanonicalHostname`")
+				return false, nil
+			}
 
-		if err := r.Client.Delete(context.Background(), fakeRoute); err != nil {
-			return ctx, quay, err
-		}
+			qctx.SupportsRoutes = true
+			prefix := fmt.Sprintf("router-%s.", rt.Status.Ingress[0].RouterName)
+			cnname := rt.Status.Ingress[0].RouterCanonicalHostname
+			qctx.ClusterHostname = strings.TrimPrefix(cnname, prefix)
+			r.Log.Info("Detected cluster hostname " + qctx.ClusterHostname)
+			return true, nil
+		},
+	); err != nil {
+		return err
+	}
 
-		return ctx, quay, nil
+	if qctx.ServerHostname == "" {
+		qctx.ServerHostname = fmt.Sprintf(
+			"%s-quay-%s.%s",
+			quay.GetName(),
+			quay.GetNamespace(),
+			qctx.ClusterHostname,
+		)
 	}
 
-	r.Log.Info("cluster does not support `Route` API", "error", err)
+	wildcard, err := getCertificatesPEM(rt.Spec.Host + ":443")
+	if err != nil {
+		return err
+	}
+	qctx.ClusterWildcardCert = wildcard
 
-	return ctx, quay, nil
+	return r.Client.Delete(ctx, &rt)
 }
 
-func (r *QuayRegistryReconciler) checkObjectBucketClaimsAvailable(ctx *quaycontext.QuayRegistryContext, quay *v1.QuayRegistry, configBundle map[string][]byte) (*quaycontext.QuayRegistryContext, *v1.QuayRegistry, error) {
-	datastoreName := types.NamespacedName{Namespace: quay.GetNamespace(), Name: quay.GetName() + "-quay-datastore"}
-	var objectBucketClaims objectbucket.ObjectBucketClaimList
-	if err := r.Client.List(context.Background(), &objectBucketClaims); err == nil {
-		r.Log.Info("cluster supports `ObjectBucketClaims` API")
-
-		ctx.SupportsObjectStorage = true
-
-		found := false
-		for _, obc := range objectBucketClaims.Items {
-			if obc.GetNamespace()+"/"+obc.GetName() == datastoreName.String() {
-				found = true
-				r.Log.Info("`ObjectBucketClaim` exists")
-
-				var datastoreSecret corev1.Secret
-				if err = r.Client.Get(context.Background(), datastoreName, &datastoreSecret); err != nil {
-					r.Log.Error(err, "unable to retrieve Quay datastore `Secret`")
+func (r *QuayRegistryReconciler) checkObjectBucketClaimsAvailable(
+	ctx context.Context, qctx *quaycontext.QuayRegistryContext, quay *v1.QuayRegistry,
+) error {
+	dstorensn := types.NamespacedName{
+		Name:      fmt.Sprintf("%s-quay-datastore", quay.GetName()),
+		Namespace: quay.GetNamespace(),
+	}
 
-					return ctx, quay, err
-				}
+	var claims objectbucket.ObjectBucketClaimList
+	if err := r.Client.List(ctx, &claims); err != nil {
+		r.Log.Info("cluster does not support `ObjectBucketClaim` API")
+		return nil
```

Comment:
This feels like an error, but the message is logged as info, and no error is returned. What am I missing?

Reply:
I kept this as is to maintain the behavior already in place. That being said I think it makes sense to return an error here (this will make the reconcile loop to fail and a proper status to be recorded). I will fix it.